File: 백준/Gold/17281. ⚾/⚾.cc
#include <iostream>
#include <algorithm>
using namespace std;

int innings;
int predictions[50][9];
int hit_order[9];
int hit[9];
int ans;
int cur_idx;

int calc() {
    int point = 0;
    cur_idx = 0;
    for (int i = 0; i < innings; i++) {
        int out_count = 0;
        int runners[3] = {0, 0, 0};
        while (out_count < 3) {
            int cur_hitter = predictions[i][hit_order[cur_idx]];
            switch (cur_hitter) {
                case 0:
                    out_count++;
                    break;
                case 1:
                    if (runners[2]) {
                        point++;
                        runners[2] = 0;
                    }
                    if (runners[1]) {
                        runners[2] = 1;
                        runners[1] = 0;
                    }
                    if (runners[0]) runners[1] = 1;
                    runners[0] = 1;
                    break;
                case 2:
                    if (runners[2]) {
                        point++;
                        runners[2] = 0;
                    }
                    if (runners[1]) point++;
                    if (runners[0]) runners[2] = 1;
                    runners[1] = 1;
                    runners[0] = 0;
                    break;
                case 3:
                    if (runners[2]) point++;
                    if (runners[1]) point++;
                    if (runners[0]) point++;
                    runners[0] = 0;
                    runners[1] = 0;
                    runners[2] = 1;
                    break;
                case 4:
                    if (runners[2]) point++;
                    if (runners[1]) point++;
                    if (runners[0]) point++;
                    point++;
                    runners[0] = 0;
                    runners[1] = 0;
                    runners[2] = 0;
                    break;
            }
            cur_idx = (cur_idx + 1) % 9;
        }
    }
    return point;
}

void solve(int idx) {
    if (idx == 3) {
        hit_order[idx] = 0;
        solve(idx + 1);
    }
    if (idx == 9) {
        cur_idx = 0;
        ans = max(ans, calc());
        return;
    }
    for (int i = 1; i < 9; i++) {
        if (hit[i]) continue;
        hit[i] = 1;
        hit_order[idx] = i;
        solve(idx + 1);
        hit[i] = 0;
    }
}

int main(void) {
    cin >> innings;
    for (int i = 0; i < innings; i++) {
        for (int j = 0; j < 9; j++) {
            cin >> predictions[i][j];
        }
    }
    solve(0);
    printf("%d\n", ans);
}

File: 백준/Gold/17281. ⚾/⚾.py

def calc_score():
    cur_idx = 0
    score = 0
    for i in range(N):
        if score + best_left[i] < ans:
            return 0
        out_count = 0
        base1 = 0
        base2 = 0
        base3 = 0
        while out_count < 3:
            hit = matrix[i][buf[cur_idx]]
            if hit == 0:
                out_count += 1
            elif hit == 1:
                score += base3
                base3 = base2
                base2 = base1
                base1 = 1
            elif hit == 2:
                score += base3 + base2
                base3 = base1
                base2 = 1
                base1 = 0
            elif hit == 3:
                score += base3 + base2 + base1
                base3 = 1
                base2 = 0
                base1 = 0
            else:
                score += base3 + base2 + base1 + 1
                base3 = 0
                base2 = 0
                base1 = 0
            cur_idx = (cur_idx + 1) % 9
    return score


def solve(cnt):
    global ans
    if ans == best_left[0]:
        return
    if cnt == 9:
        ans = max(ans, calc_score())
        return
    if cnt == 3:
        buf[cnt] = 0
        solve(cnt + 1)
        return
    for i in range(1, 9):
        if visited[i]: continue
        buf[cnt] = i
        visited[i] = 1
        solve(cnt + 1)
        visited[i] = 0


N = int(input())
matrix = [list(map(int, input().split())) for _ in range(N)]
best = []
for row in matrix:
    outs = row.count(0)
    if outs == 1:
        best.append(24)
    elif outs == 2:
        best.append(14)
    else:
        best.append(9 - outs)
best_left = [0] * (N + 1)
for i in range(N-1, -1, -1):
    best_left[i] = best_left[i+1] + best[i]
buf = [0] * 9
visited = [0] * 9
ans = 0
solve(0)
print(ans)
